src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -323,3 +500,237 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
+                }
+
+                accumulated_count += count;
+            }
+            MPL_free(src_ptr);
+            MPL_free(typerep_vec);
+        }
+    } else {
+        if ((HANDLE_IS_BUILTIN(acc_op))
+            && ((*MPIR_OP_HDL_TO_DTYPE_FN(acc_op)) (source_dtp) == MPI_SUCCESS)) {
+            /* get the function by indexing into the op table */
+            uop = MPIR_OP_HDL_TO_FN(acc_op);
+        }
+        void *save_targetbuf = NULL;
+        void *host_targetbuf = NULL;
+
+        host_targetbuf = MPIR_gpu_host_swap(target_buf, target_count, target_dtp);
+
+        if (host_targetbuf) {
+            save_targetbuf = target_buf;
+            target_buf = host_targetbuf;
+        }
+
+        if (is_empty_source == TRUE || HANDLE_IS_BUILTIN(target_dtp)) {
+            /* directly apply op if target dtp is predefined dtp OR source buffer is empty */
+            MPI_Aint tmp_count = source_count;
+            (*uop) (source_buf, target_buf, &tmp_count, &source_dtp);
+        } else {
+            i = 0;
+            curr_loc = typerep_vec[0].iov_base;
+            curr_len = typerep_vec[0].iov_len;
+            accumulated_count = 0;
+            while (i != vec_len) {
+                if (curr_len < type_size) {
+                    MPIR_Assert(i != vec_len);
+                    i++;
+                    curr_len += typerep_vec[i].iov_len;
+                    continue;
+                }
+
+                MPI_Aint count;
+                MPIR_Assign_trunc(count, curr_len / type_size, MPI_Aint);
+
+                if (src_ptr) {
+                    MPI_Aint unpacked_size;
+                    MPIR_Typerep_unpack((char *) source_buf + src_type_stride * accumulated_count,
+                                        source_dtp_size, src_ptr, 1, source_dtp, 0, &unpacked_size);
+                    (*uop) (src_ptr, (char *) target_buf + MPIR_Ptr_to_aint(curr_loc), &count,
+                            &type);
+                } else {
+                    (*uop) ((char *) source_buf + src_type_stride * accumulated_count,
+                            (char *) target_buf + MPIR_Ptr_to_aint(curr_loc), &count, &type);
+                }
+
+                if (curr_len % type_size == 0) {
+                    i++;
+                    if (i != vec_len) {
+                        curr_loc = typerep_vec[i].iov_base;
+                        curr_len = typerep_vec[i].iov_len;
+                    }
+                } else {
+                    curr_loc = (void *) ((char *) curr_loc + type_extent * count);
+                    curr_len -= type_size * count;
+                }
+
+                accumulated_count += count;
+            }
+
+            MPL_free(src_ptr);
+            MPL_free(typerep_vec);
+        }
+
+        if (save_targetbuf) {
+            MPIR_gpu_swap_back(target_buf, save_targetbuf, target_count, target_dtp);
+            target_buf = save_targetbuf;
+        }
+    }
```

Comment:
This function is too complex to read -- I understand it is nothing of your fault :) . So at a high level ---
```
condition = ...
If (condition == condition_1) 
    op_pack
elif (condition == condition_2)
    op_unpack
else
    op_fallback
```
Do I understand correctly? If so, can we refactor accordingly?

Reply:
I thought we discussed that the refactoring of this function will be done in a separate PR.

Reply:
I guess we did. But I can't really review the function as is. How about let me refactor it over the weekend?